### PR TITLE
Add AJA adapter

### PR DESCRIPTION
**Pull request by sambaiz**
New adapter for AJA

### CODE REVIEW

File: usersync/usersyncers/syncer.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package usersyncers
 
 import (
+	"github.com/prebid/prebid-server/adapters/aja"
```
[the remark this answers is not in the document]
Comment (sambaiz) [reply usersync/usersyncers/syncer.go]: Fixed, thanks.

File: exchange/adapter_map.go [hunk 2]
```diff
@@ -2,6 +2,7 @@ package exchange
 
 import (
 	"fmt"
+	"github.com/prebid/prebid-server/adapters/aja"
```
[the remark this answers is not in the document]
Comment (sambaiz) [reply exchange/adapter_map.go]: Fixed

File: adapters/aja/aja.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	// split imps by tagid
+	impsByTagID := map[string][]openrtb.Imp{}
+	for _, imp := range bidReq.Imp {
+		extAJA, err := parseExtAJA(imp)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		imp.TagID = extAJA.AdSpotID
+		imp.Ext = nil
```
[the remark this answers is not in the document]
Comment (sambaiz) [reply adapters/aja/aja.go:27]: I think the `imp` is a new variable assigned by for-range so these operation has no effect to others. What do you think?

[the remark this answers is not in the document]
Comment (guscarreon) [reply adapters/aja/aja.go:28]: You are totally right. I double-checked and we'd only get in trouble irf we dereferenced the elements of the `Imp` slice like: `bidReq.Imp[i]`. Will mark this comment as resolved.

File: adapters/aja/aja.go [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+		if _, ok := impsByTagID[imp.TagID]; !ok {
+			tagIDs = append(tagIDs, imp.TagID)
+		}
+		impsByTagID[imp.TagID] = append(impsByTagID[imp.TagID], imp)
+	}
+
+	req := *bidReq
+	for _, tagID := range tagIDs {
+		req.Imp = impsByTagID[tagID]
+		body, err := json.Marshal(req)
```
Comment (guscarreon) [inline adapters/aja/aja.go:38]: Do we really need `tagIDs := []string{}`? I believe the following changes use less memory and are equivalent. Am I wrong? Also, `bidReq` is already a shallow copy where is safe to modify `bidReq.Imp`.
```
17   func (a *AJAAdapter) MakeRequests(bidReq *openrtb.BidRequest, extraInfo *adapters.ExtraRequestInfo) (adapterReqs []*adapters.RequestData, errs []error) {
18       // split imps by tagid
19 -     tagIDs := []string{}
20       impsByTagID := map[string][]openrtb.Imp{}
21       for _, imp := range bidReq.Imp {
22           extAJA, err := parseExtAJA(imp)
23           if err != nil {
24               errs = append(errs, err)
25               continue
26           }
27           imp.TagID = extAJA.AdSpotID
28           imp.Ext = nil
29 -         if _, ok := impsByTagID[imp.TagID]; !ok {
30 -             tagIDs = append(tagIDs, imp.TagID)
31 -         }
32           impsByTagID[imp.TagID] = append(impsByTagID[imp.TagID], imp)
33       }
34  
35 -     req := *bidReq
36 -     for _, tagID := range tagIDs {
37 -         req.Imp = impsByTagID[tagID]
38 -         body, err := json.Marshal(req)
   +     for _, imps := range impsByTagID {
   +         bidReq.Imp = imps
   +         body, err := json.Marshal(bidReq)
39           if err != nil {
40               errs = append(errs, err)
41               continue
42           }
43           adapterReqs = append(adapterReqs, &adapters.RequestData{
44               Method: "POST",
45               Uri:    a.endpoint,
46               Body:   body,
47           })
48       }
49  
50       return
51   }
adapters/aja/aja.go
```

File: adapters/aja/aja.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		}}
+	}
+
+	bidderResp := adapters.NewBidderResponseWithBidsCapacity(len(bidReq.Imp))
+	var errors []error
+
+	for _, seatbid := range bidResp.SeatBid {
+		for _, bid := range seatbid.Bid {
+		impID:
+			for _, imp := range bidReq.Imp {
```
[the remark this answers is not in the document]
Comment (sambaiz) [reply adapters/aja/aja.go:101]: No, this label is unnecessary. I have removed.

File: adapters/aja/aja.go (hunk 4 above)
Comment (sambaiz) [reply adapters/aja/aja.go:38]: Yes, I believe it needs. `range impsByTagID` order is changed with each execution and this cause to make tests sometimes failure so it need to fix order with `tagIDs` slice.

> bidReq is already a shallow copy 

Where is it copied? If this is removed, the test is failed.

**Comment by sambaiz** [comment]
I have increased coverage.

```
$ go tool cover -func=cover.out 
github.com/prebid/prebid-server/adapters/aja/aja.go:17:         MakeRequests    90.5%
github.com/prebid/prebid-server/adapters/aja/aja.go:55:         parseExtAJA     100.0%
github.com/prebid/prebid-server/adapters/aja/aja.go:76:         MakeBids        96.0%
github.com/prebid/prebid-server/adapters/aja/aja.go:128:        NewAJABidder    100.0%
github.com/prebid/prebid-server/adapters/aja/usersync.go:10:    NewAJASyncer    100.0%
total:                                                          (statements)    94.4%
```

File: adapters/aja/aja.go (hunk 4 above)
Comment (guscarreon) [reply adapters/aja/aja.go:38]: @sambaiz, thanks for the feedback. We were wrong, modifying `bidReq.Imp` at this level does change the original request. Thanks for pointing this out

File: config/config.go [hunk 6]
```diff
@@ -689,6 +690,7 @@ func SetupViper(v *viper.Viper, filename string) {
 	v.SetDefault("adapters.adpone.endpoint", "http://rtb.adpone.com/bid-request?src=prebid_server")
 	v.SetDefault("adapters.adtelligent.endpoint", "http://hb.adtelligent.com/auction")
 	v.SetDefault("adapters.advangelists.endpoint", "http://nep.advangelists.com/xp/get?pubid={{.PublisherID}}")
+	v.SetDefault("adapters.aja.endpoint", "https://ad.as.amanad.adtdp.com/v1/bid/4")
```
Comment (SyntaxNode) [inline config/config.go:693]: I'd just like to verify that the `4` is not specific to AppNexus/Xandr or another specific host.

Comment (sambaiz) [reply config/config.go:693]: Yes. This `4` is not for specific host but for general prebid-server.

**Pull request merged**